### PR TITLE
Phi2 rewrite

### PR DESCRIPTION
**Pull request by winglian**
- start over with latest phi2 modeling code from hf repo
- Gradient checkpointing for ParallelBlock by @abacaj via https://fxtwitter.com/abacaj/status/1743749460872634506
- update gradient checkpointing signature to match transformers
- Properly enable flash attention, thanks @vikhyat via https://fxtwitter.com/vikhyatk/status/1743876177037861291
- ~~disable most of the upcasting to float32 in favor of bfloat16~~

### CODE REVIEW

**Comment by casper-hansen** [comment]
Looks good to me. I would remove the code you commented out. You can come back to the PR if you need to look up what changed.

Also, what are the possibilities of using sample packing with Phi2?

**Comment by winglian** [comment]
@casper-hansen will clean up the commented out code

As far as sample packing, it should be pretty straightforward. I started working on a fix for the previous implementation https://github.com/OpenAccess-AI-Collective/axolotl/pull/877 but I may simply start over. 

You had mentioned last year figuring out a way to manage sample packing across all the architectures in a more manageable way. I'm happy to take a stab at it if you have a prrof of concept or anything.

**Comment by casper-hansen** [comment]
> You had mentioned last year figuring out a way to manage sample packing across all the architectures in a more manageable way. I'm happy to take a stab at it if you have a prrof of concept or anything.

I had a branch going but didn't get to test and further implement it as I got busy with other stuff. The concept is to have one implementation that can be managed more easily managed in one module.

https://github.com/OpenAccess-AI-Collective/axolotl/tree/refactor-flash-attention

File: src/axolotl/core/trainer_builder.py [hunk 1]
```diff
@@ -843,7 +844,14 @@ def build_collator(self, training_args: AxolotlTrainingArguments, **kwargs):
         if self.cfg.model_config_type == "mamba":
             return MambaDataCollator(tokenizer=self.tokenizer)
 
-        return BatchSamplerDataCollatorForSeq2Seq(
+        if training_args.sample_packing:
```
Comment (NanoCode012) [inline src/axolotl/core/trainer_builder.py:847]: I would recommend maybe consolidate the class?

```py
data_collator = BatchSamplerDataCollatorForSeq2Seq if training_args.sample_packing else DataCollatorForSeq2Seq
```

Comment (winglian) [reply src/axolotl/core/trainer_builder.py:847]: Screenshot 2024-01-08 at 12 11 10 PM

my IDE doesn't like that 😭

**Comment by winglian** [comment]
alright, looks good on a single 4090 https://api.wandb.ai/links/oaaic/51qvcv4z

**Comment by fakerybakery** [comment]
Hi, do I need to change any configuration options or just use the default ones w/ Phi 2?

**Pull request merged**